documentation/src/docs/wire-protocol.md

```diff
@@ -7,106 +7,103 @@ You may obtain a copy of the License at
 
-The remainder of the fields are specific to the type of message. A few important messages are listed below.
+- **Message Type**: An Integer (4 bytes) identifies the message type and determines the subsequent fields. (Note that the protocol can be extended by adding new types.)
+- **Length**:  An Integer (4 bytes) (Messages should be less than 2<sup>24</sup>, but the upper bits remain zero). Payload size of the message (possibly zero, indicating there is no data). The remainder of the fields are specific to the type of the message. A few important messages are listed below.
 
-General
--------
+# General
 
-### Partial Message - Request/Reply
+## KeepAlive - Request/Reply
 
-1.  Begin/Middle/End - Enum (1 byte)
-2.  Data
-A partial message is one that was broken up when being sent over the wire. (For any reason). The whole message is reconstructed by reading the partial messages in sequence and assembling them into a whole. It is not valid to attempt to start a new partial message before completing the previous one.
+# Reading
 
-### KeepAlive - Request/Reply
+## Read Segment - Request
 
-1.  Data - uninterpreted data of the length of the message. (Usually 0 bytes)
+1. `Segment`: String (2 bytes) followed by that many bytes of Java's Modified UTF-8. This Segment indicates the Stream Segment that was read.
+2. `Offset`: Long (8 bytes). The offset in the Stream Segment to read from.
+3. `Reply`: Integer (4 bytes). The clients can request for the required length to the server (but the server may allot a different number of bytes).
+4. `delegationToken`: String (2 byte) followed by that many bytes of Java's Modified UTF-8. This was added to perform _auth_. It is an opaque-to-the-client token provided by the Controller that says it's allowed to make this call.
 
-Reading
--------
 
-### Read Segment - Request
+## Segment Read - Reply
 
-1.  Segment to read - String (2 byte length, followed by that many bytes of Java's Modified UTF-8)
-2.  Offset to read from - Long (8 bytes)
-3.  Suggested Length of Reply - int (4 bytes)
-    1.  This is how much data the client wants. They won't necessarily get that much.
+1. `Segment`: String (2 bytes) followed by that many bytes of Java's Modified UTF-8). This Segment indicates the Stream Segment that was read.
+2. `Offset`: Long (8 bytes). The offset in the Stream Segment to read from.
+3. `Tail`: Boolean (1 bit). If the read was performed at the tail of the Stream Segment.
+4. `EndOfSegment`: Boolean (1 bit). If the read was performed at the end of the Stream Segment.
+5. `Data`: Binary (remaining length in the message).
 
-### Segment Read - Reply
+The client requests to read from a particular Segment at a particular Offset. It then receives one or more replies in the form of `SegmentRead` messages. These contain the data they requested (assuming it exists). The server may decide transferring to the client more or less data than it was asked for, splitting that data in a suitable number of reply messages.
```

Comment:
For consistency, maybe "Offset" should be in `this` format in this paragraph.

Reply:
@RaulGracia  Accepted suggestion chnaged to `Offset`